share/functions/fish_clipboard_paste.fish

```diff
@@ -21,6 +21,12 @@ function fish_clipboard_paste
     # otherwise the output looks really confusing.
     set data (string split \r -- $data)
 
+    if not isatty stdout
+        # If we're redirected, just write the data
+        printf %s\n $data
```

Comment:
I used to have this as part of my config but it broke my integration with tmux and neovim. (Instead of rewriting all my clipboard handling each time for each application that wants to integrate, I just farm out the clipboard operation to fish.)

The problem is that if you forcibly insert line breaks, some applications won't strip them out on paste (at least by default). I wanted to be able to copy a word to the clipboard in neovim/tmux and then paste it into another doc in another terminal instance, but emitting a trailing `\n` causes a hard line break to be inserted everywhere I paste.

I've since modified my clipboard wrappers to omit any additional line breaks and I've found that to be a much more zero-stress solution. If you _do_ want a line break after pasting, you can always just hit enter/ctrl-j, but adding a line break can cause problems. We already handle commands that emit data without a trailing `\n` very nicely so there's no harm (to us) in excluding that.

Oh yeah, I remembered another issue I ran into: you also especially don't want to force either execution of the clipboard contents or forcibly starting a second line if the user is using `fish_clipboard_paste` to paste into another shell.

Reply:
I agree; I use a similar script for copying/pasting text on any OS and I also made it not produce newlines out of thin air.
That's important when using it to paste into a text editor.

Reply:
Should it not "produce newlines out of thin air", or have trailing newlines at all?

Those are two different things.

Reply:
We should print the clipboard contents verbatim. I realized that we also need to make sure that we don't drop actual newlines. This should do it but it requires GNU sed AFAIK

```diff
diff --git a/share/functions/fish_clipboard_paste.fish b/share/functions/fish_clipboard_paste.fish
index 49a2dc1be..5509154ec 100644
--- a/share/functions/fish_clipboard_paste.fish
+++ b/share/functions/fish_clipboard_paste.fish
@@ -1,19 +1,19 @@
 function fish_clipboard_paste
     set -l data
     if type -q pbpaste
-        set data (pbpaste 2>/dev/null)
+        set data "$(pbpaste 2>/dev/null)E"
     else if set -q WAYLAND_DISPLAY; and type -q wl-paste
-        set data (wl-paste 2>/dev/null)
+        set data "$(wl-paste 2>/dev/null)E"
     else if set -q DISPLAY; and type -q xsel
-        set data (xsel --clipboard)
+        set data "$(xsel --clipboard)E"
     else if set -q DISPLAY; and type -q xclip
         set data (xclip -selection clipboard -o 2>/dev/null)
     else if type -q powershell.exe
-        set data (powershell.exe Get-Clipboard | string trim -r -c \r)
+        set data "$(powershell.exe Get-Clipboard | string trim -r -c \r)E"
     end
 
     # Issue 6254: Handle zero-length clipboard content
-    if not string match -qr . -- "$data"
+    if not string match -qr .E -- "$data"
         return 1
     end
 
@@ -23,10 +23,12 @@ function fish_clipboard_paste
 
     if not isatty stdout
         # If we're redirected, just write the data
-        printf %s\n $data
+        printf %s $data | sed '$s/E$//'
         return
     end
 
+    set data (printf %s $data | sed '$s/E$//')
+
     # If the current token has an unmatched single-quote,
     # escape all single-quotes (and backslashes) in the paste,
     # in order to turn it into a single literal token.
```

Reply:
>set data "$(powershell.exe Get-Clipboard | string trim -r -c \r)E"

You want `string collect -N`. No fiddling with `sed` or awkward special-insert chars. Just:

```fish
set data (xsel --clipboard | string collect -N)
```

Reply:
The default behavior on all platforms I have observed with clipboard functionality is that you can copy a single word (without including/selecting the line break) and then inspect the system clipboard contents as binary and observe that no trailing new line has been added. You should be able to select a word in the middle of a line, copy it, then select it and paste over it without any whitespace changes and without the client needing to intercept and eliminate trailing whitespace.

We don't have an issue with that *in fish* because both presence and absence of a trailing newline is coalesced in any subshell (i.e. `printf "%s" (echo foo)` and `printf "%s" (echo -n foo)` are identical). But when we touch the system clipboard, the ultimate target may or may not be fish and may or may not preserve trailing line breaks.

Reply:
5bbe0073e258cf1ad22d48f79ac8b5dca874ea00 should do it.

Reply:
so we can just use `string collect -N` here as well to do the same thing